File: c/2-motor-controller/main.c
#include <bits/types/siginfo_t.h>
#include <fcntl.h>
#include <math.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/ioctl.h>
#include <unistd.h>

#include <i2c/smbus.h>
#include <linux/i2c-dev.h>
#include <pigpio.h>

#define I2C_ADAPTER_NUMBER "1"
const char I2C_ADAPTER_PATH[] = "/dev/i2c-" I2C_ADAPTER_NUMBER;
const uint32_t ADS7830_ADDRESS = 0x48;
const uint32_t MOTOR_LINE_NUMBER = 13;

const uint64_t PERIOD_MS = 5000;
const uint64_t PWM_CHANGES = 50;
const uint64_t PWM_MIN = 0.2 * PI_HW_PWM_RANGE;
const uint64_t PWM_MAX = 1.0 * PI_HW_PWM_RANGE;
const uint64_t PWM_FREQUENCY = 1000;
const uint64_t SLEEP_DURATION_NS = PERIOD_MS * 1000 / PWM_CHANGES;

bool do_continue = true;

void interrupt_handler(int) {
  printf("\nGracefully stopping\n");
  do_continue = false;
}

int32_t read_potentiometer_value(int i2c_file) {
  if (i2c_smbus_write_byte(i2c_file, 0x84) < 0) {
    perror("writing i2c ADC command failed\n");
    return -1;
  }

  int32_t value = i2c_smbus_read_byte(i2c_file);
  if (value < 0) {
    perror("reading i2c ADC value failed\n");
    return -1;
  }

  return value;
}

int main(int, char **) {
  if (gpioInitialise() < 0) {
    perror("pigpio initialization failed\n");
    goto end;
  }

  if (gpioSetSignalFunc(SIGINT, interrupt_handler)) {
    perror("Setting sigaction failed\n");
    goto end;
  }

  printf("%s", I2C_ADAPTER_PATH);
  int i2c_file = open(I2C_ADAPTER_PATH, O_RDWR);
  if (i2c_file < 0) {
    perror("opening i2c adapter failed\n");
    goto end;
  }

  if (ioctl(i2c_file, I2C_SLAVE, ADS7830_ADDRESS) < 0) {
    perror("assigning i2c adapter address failed\n");
    goto end;
  }

  printf("Controlling motor from C.\n");

  while (do_continue) {
    for (size_t i = 0; i < PWM_CHANGES; ++i) {
      int32_t value = read_potentiometer_value(i2c_file);
      if (value < 0)
        continue;

      printf("selected duty cycle: %.2f\n", (double)value / UINT8_MAX);

      // const double_t sin_value = sin((double_t)i / PWM_CHANGES * 2 * M_PI);
      // const double_t ratio = (sin_value + 1.) / 2.;
      const uint64_t duty_cycle =
          PWM_MIN + (PWM_MAX - PWM_MIN) * value / UINT8_MAX;
      if (gpioHardwarePWM(MOTOR_LINE_NUMBER, PWM_FREQUENCY, duty_cycle) < 0) {
        perror("Setting PWM failed\n");
        goto close_pigpio;
      }

      if (!do_continue)
        break;
      usleep(SLEEP_DURATION_NS);
    }
  }

  if (gpioHardwarePWM(MOTOR_LINE_NUMBER, 0, 0)) {
    perror("Disabling PWM failed\n");
  }

close_pigpio:
  gpioTerminate();
end:
  return EXIT_SUCCESS;
}

File: zig/2-motor-controller/main.zig
const std = @import("std");
const pwm = @import("pwm");

const c = @cImport({
    @cInclude("signal.h");
    @cInclude("linux/i2c-dev.h");
    @cInclude("i2c/smbus.h");
    @cInclude("sys/ioctl.h");
});

const i2c_adapter_number = 1;
const i2c_adapter_path = std.fmt.comptimePrint("/dev/i2c-{}", .{i2c_adapter_number});
const ads7830_address: c_int = 0x48;
const motor_pwm_channel: u8 = 1; // gpio 13

const period_ms: u64 = 5000;
const pwm_updates_per_period: u64 = 50;
const pwm_min: f32 = 0.2;
const pwm_max: f32 = 1.0;
const pwm_frequency: f32 = 1000;
const sleep_time_ns: u64 = period_ms * std.time.ns_per_ms / pwm_updates_per_period;

var do_continue = true;
pub fn interrupt_handler(_: c_int) callconv(.C) void {
    std.debug.print("\nGracefully stopping\n", .{});
    do_continue = false;
}
const interrupt_sigaction = c.struct_sigaction{
    .__sigaction_handler = .{ .sa_handler = &interrupt_handler },
};

pub fn main() !void {
    if (c.sigaction(c.SIGINT, &interrupt_sigaction, null) != 0) {
        return error.SigactionNotSet;
    }

    var chip = try pwm.Chip.init(0);
    defer chip.deinit();

    const i2c_file = try std.fs.openFileAbsolute(
        i2c_adapter_path,
        std.fs.File.OpenFlags{ .mode = .read_write },
    );
    defer i2c_file.close();

    if (c.ioctl(i2c_file.handle, c.I2C_SLAVE, ads7830_address) < 0) {
        return error.SettingI2cSlave;
    }

    std.debug.print("Controlling motor from Zig.\n", .{});

    var channel = try chip.channel(motor_pwm_channel);
    defer channel.deinit();

    try channel.setParameters(.{
        .frequency = pwm_frequency,
        .duty_cycle_ratio = 0,
    });
    try channel.enable();

    main_loop: while (true) {
        for (0..pwm_updates_per_period) |i| {
            _ = c.i2c_smbus_write_byte(i2c_file.handle, 0x84);
            const value = c.i2c_smbus_read_byte(i2c_file.handle);
            std.debug.print("selected duty cycle: {}", .{value});

            updateDutyCycle(channel, i) catch |err| {
                std.debug.print("error updating duty cycle: {}", .{err});
            };

            if (!do_continue) break :main_loop;
            std.time.sleep(sleep_time_ns);
        }
    }
}
fn updateDutyCycle(channel: *pwm.Channel, step: u64) !void {
    const sin_arg_ratio = @as(f32, @floatFromInt(step)) / pwm_updates_per_period;
    const sin = std.math.sin(sin_arg_ratio * 2 * std.math.pi);
    const ratio = (sin + 1) / 2;
    const duty_cycle = pwm_min + ratio * (pwm_max - pwm_min);
    try channel.setParameters(.{
        .frequency = null,
        .duty_cycle_ratio = duty_cycle,
    });
}
